PostgreSQL: Add initial v17.4 package

hgy59 opened the pull request:
## Description

This PR adds the PostgreSQL package with significant improvements to security, architecture support, and adds the PostGIS spatial extension.

_Motivation:_  This is a followup of the PR #3826 (fork was deleted)

### PostGIS Spatial Extension
- Added **PostGIS 3.5.2** with GEOS 3.12.3 and PROJ support
- Enables geographic/geometric data types and spatial queries
- Cross-compilation support for multiple architectures
- **PostGIS is only included when GCC >= 5.0 is available (DSM 7+)**

### Security Improvements
- Changed default authentication from `md5` to `scram-sha-256`
- Added password complexity validation in install wizard
- Improved backup/restore handling during upgrades

### Architecture Support
- **Supported with PostGIS**: x64, aarch64, evansport, armv7 (generic) on DSM 7.0+ (GCC 5+)
- **Supported without PostGIS**: comcerto2k (DSM 7.0/7.1), DSM 6.2.4 (GCC 4.9)
- **Excluded**: ARMv5 (88f6281) - libicu-74.2 doesn't support it

## Dependencies Added

- `cross/postgis` - PostGIS 3.5.2
- `cross/libgeos` - GEOS 3.12.3
- `cross/libproj` - PROJ cartographic library
- `native/sqlite` - Required for PROJ build

## Checklist

- [x] Build rule `all-supported` completed successfully
- [x] New installation of package completed successfully
- [x] Package upgrade completed successfully (Manually install the package again)
- [x] Package [functionality was tested](https://github.com/SynoCommunity/spksrc/wiki/Package-Update-Policy#tests-checks)
- [x] Any needed [documentation](https://github.com/SynoCommunity/spksrc/wiki/Create-documentation) is updated/created


### Type of change

<!--Please use any relevant tags.-->
- [x] New Package
- [x] This change requires a documentation update (e.g. Wiki)

DigitalBox98 committed:
PostgreSQL 11.6 including wizard install/uninstall

DigitalBox98 committed:
Enhancements for PostgreSQL :
- Dump location included in the wizard
- When upgrading, user/password/port is retrieved
- Status daemon fixed by using old-style dsmcontrol

hgy59 committed:
update to postgresql v11.10

- add patch to fix linking with ncursesw
- move wizard files to src/wizard
- fix lint result for src/wizard/uninstall_uifile

hgy59 committed:
update pgsql and migrate to generic service installer

- update postgres to v11.15 and make compatible with DSM6 and DSM7
- use custom EFF_USER as "sc-" prefix is invalid for postgresql
- use custom privilege file for DSM<7 to install/uninstall/upgrade as root but run as package user
- adjust privileges for DSM<7

hgy59 commented:
@DigitalBox98 @ymartin59 are you still interested in PostgreSQL package?
It is now working on DSM 7 and DSM 6.

hgy59 committed:
udpate to v11.16

RMHogervorst commented:
I can't get this build for alpine. Am I missing someting? 

I tried:
- `make spk-postgres-alpine-7.0` from the rootfolder
- moving into the postgres folder and running `make alpine-7.0`
- idem with `make syno-alpine-7.0`
result: `make: *** No rule to make target 'syno-alpine-7.0'.  Stop.`

hgy59 commented:
@RMHogervorst did you try (in the spk/postgres folder)?
```
make arch-alpine-7.0
```

an alternative would be:
```
make ARCH=alpine TCVERSION=7.0
```

RMHogervorst commented:
wow, that was the piece of missing information! I think I missed this because the developer docs talk about `make arch-88f6281` and I read that as make [architecturestring], but the arch part is separate!

It now fails on zlib
```
Downloading files for zlib
===>    wget https://zlib.net/zlib-1.2.11.tar.gz
https://zlib.net/zlib-1.2.11.tar.gz:
2023-01-09 06:26:32 ERROR 404: Not Found.
```

RMHogervorst commented:
it seems the current version of zlib is 1.2.13 (released October 13, 2022 ) because of a cve all previous versions are unavailable I think. 

Here are the changes, not in an PR because I think you should check the hash digests yourself since I am a random person on the internet.

cross/zlib/makefile
```
PKG_NAME = zlib
PKG_VERS = 1.2.13
PKG_EXT = tar.gz
PKG_DIST_NAME = $(PKG_NAME)-$(PKG_VERS).$(PKG_EXT)
PKG_DIST_SITE = https://zlib.net
PKG_DIR = $(PKG_NAME)-$(PKG_VERS)

DEPENDS =

HOMEPAGE = https://zlib.net/
COMMENT  = A Massively Spiffy Yet Delicately Unobtrusive Compression Library.
LICENSE  = zlib-license

ADDITIONAL_CFLAGS = -Os

include ../../mk/spksrc.cross-cc.mk
```

and the digests in cross/zlib/digests
```
zlib-1.2.13.tar.gz SHA1 55eaa84906f31ac20d725aa26cd20839196b6ba6
zlib-1.2.13.tar.gz SHA256 b3a24de97a8fdbc835b9833169501030b8977031bcb54b3b3ac13740f846ab30
zlib-1.2.13.tar.gz MD5 9b8aa094c4e5765dabf4da391f00d15c
```

hgy59 commented:
@RMHogervorst we already have the current zlib on the master branch. So I have rebased this PR.

BTW: for alpine you can use the armv7 package created by the github build action (when it is successfully completed, and within 90 days).

RMHogervorst commented:
I feel really stupid, where can I find the installation docs? I get `invalid file format. contact developers`
I tried manual install within DSM with the zipfile for arm v7 and arm v7-6.1 but also the file inside.

hgy59 commented:
you must install *.spk files (not zip)

RMHogervorst commented:
alright this is what I tried
- unpack the downloaded zip file 
- tar the file and turn it into a spk file `tar -cvzf postgresql.spk postgresql_armv7-7.0_11.16-3`
- install the spk file (did not work)

hgy59 commented:
1. download the artifacts (*.zip)
2. unpack the artifacts -> you will get the spk 
3. In DSM Package center of your Diskstation chose "Manual Install" and select the spk file to upload and install

RMHogervorst commented:
Alright this is what lives in the unpacked zipfile. Did something go wrong? because there is no spk file here.

```
postgresql_armv7-7.0_11.16-3 tree
.
├── INFO
├── PACKAGE_ICON.PNG
├── PACKAGE_ICON_256.PNG
├── WIZARD_UIFILES
│   ├── install_uifile
│   ├── uninstall_uifile
│   └── upgrade_uifile.sh
├── conf
│   ├── privilege
│   └── resource
├── package.tgz
└── scripts
    ├── functions
    ├── installer
    ├── postinst
    ├── postuninst
    ├── postupgrade
    ├── preinst
    ├── preuninst
    ├── preupgrade
    ├── service-setup
    └── start-stop-status

3 directories, 19 files
```

hgy59 commented:
I have not a clue what's going wrong.

1. navigate to "checks - build -artifacts"
2. click (e.g.) "Packages for aarch64-6.1"
3. it will download `Packages for aarch64-6.1.zip`
4. unpack the downloaded zip will extract `postgresql_aarch64-6.1_11.16-3.spk`

This should work for all archs (I have also validated the armv7-6.1 and armv7-7.0 artifacts)

The tree you showed above looks like the content of package.tgz within the spk file.

RMHogervorst commented:
Wow this is really weird, it must be a mac specific (or my computer specific) issue. When I use the standard macos build in tool to open the zip, it opens to the folder as I described, but when I use a different tool it unzips to the spk file!

Thanks for the help!

RMHogervorst commented:
ah shoot! it doesn't install because the package requires root privileges:

<img width="697" alt="Screenshot 2023-01-10 at 07 20 25" src="https://user-images.githubusercontent.com/9538040/211476586-4eb5b626-9b00-41ea-b941-2b3ddec6a6aa.png">

hgy59 committed:
avoid root privileges on DSM 7

hgy59 commented:
@RMHogervorst can you please try with the new version?
(didn't test it, but the installer should work)

hgy59 committed:
postgresql: update to v11.18

RMHogervorst commented:
There might be an issue in the long run with no superuser to do maintainence?

```
 tail postgresql.log
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [6-1] 2023-01-10 00:14:18.543 GMT [16794] WARNING:  skipping "pg_subscription" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [7-1] 2023-01-10 00:14:18.825 GMT [16794] WARNING:  skipping "pg_database" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [8-1] 2023-01-10 00:14:18.825 GMT [16794] WARNING:  skipping "pg_db_role_setting" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [9-1] 2023-01-10 00:14:18.825 GMT [16794] WARNING:  skipping "pg_tablespace" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [10-1] 2023-01-10 00:14:18.841 GMT [16794] WARNING:  skipping "pg_pltemplate" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [11-1] 2023-01-10 00:14:18.841 GMT [16794] WARNING:  skipping "pg_auth_members" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [12-1] 2023-01-10 00:14:18.842 GMT [16794] WARNING:  skipping "pg_shdepend" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [13-1] 2023-01-10 00:14:18.842 GMT [16794] WARNING:  skipping "pg_shdescription" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [14-1] 2023-01-10 00:14:18.848 GMT [16794] WARNING:  skipping "pg_replication_origin" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [15-1] 2023-01-10 00:14:18.850 GMT [16794] WARNING:  skipping "pg_shseclabel" --- only superuser can vacuum it
ash-4.4# tail postgresql.log
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [6-1] 2023-01-10 00:14:18.543 GMT [16794] WARNING:  skipping "pg_subscription" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [7-1] 2023-01-10 00:14:18.825 GMT [16794] WARNING:  skipping "pg_database" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [8-1] 2023-01-10 00:14:18.825 GMT [16794] WARNING:  skipping "pg_db_role_setting" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [9-1] 2023-01-10 00:14:18.825 GMT [16794] WARNING:  skipping "pg_tablespace" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [10-1] 2023-01-10 00:14:18.841 GMT [16794] WARNING:  skipping "pg_pltemplate" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [11-1] 2023-01-10 00:14:18.841 GMT [16794] WARNING:  skipping "pg_auth_members" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [12-1] 2023-01-10 00:14:18.842 GMT [16794] WARNING:  skipping "pg_shdepend" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [13-1] 2023-01-10 00:14:18.842 GMT [16794] WARNING:  skipping "pg_shdescription" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [14-1] 2023-01-10 00:14:18.848 GMT [16794] WARNING:  skipping "pg_replication_origin" --- only superuser can vacuum it
2023-01-10T01:14:18+01:00 netwerkhdd postgres[16794]: [15-1] 2023-01-10 00:14:18.850 GMT [16794] WARNING:  skipping "pg_shseclabel" --- only superuser can vacuum it
```

hgy59 commented:
@RMHogervorst do you know what cleanup script/task is run by postgresql?
we should disable the pgsql internal cleanup and run a scheduled task in DSM (yes, those task can run scripts with root privileges in DSM 7).

RMHogervorst commented:
Not from the top of my head, but I guess this https://www.postgresql.org/docs/14/maintenance.html
I will check the pg_cron table and dig a bit further

21 jan. 2023 20:01:39 hgy59 ***@***.***>:

> @RMHogervorst[https://github.com/RMHogervorst] do you know what cleanup script/task is run by postgresql?
> we should disable the pgsql internal cleanup and run a scheduled task in DSM (yes, those task can run scripts with root privileges in DSM 7).
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/SynoCommunity/spksrc/pull/4267#issuecomment-1399312455], or unsubscribe[https://github.com/notifications/unsubscribe-auth/ACIYT6BG2ISYVVPGMXJ2V4TWTQXBFANCNFSM4TXSLHJA].
> You are receiving this because you were mentioned.[Tracking afbeelding][https://github.com/notifications/beacon/ACIYT6BMZXWYDJCKNKUX27TWTQXBFA5CNFSM4TXSLHJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOKNT5ARY.gif]
>

RMHogervorst commented:
As far as I can see these are automatic cleanup tasks from the 'Autovacuum Daemon'. 
- autovacuum  VACUUM
- updating the statistics on tables ANALYZE

The errors only happen on the system (pg_catalog) tables. so pg_tablespace, db_role_settings, pg_database etc. 
I don't think this will cause many problems (but I'm no dba)

The settings for this deamon should be in the postgresql.conf file. 
But `autovacuum = off` in my /etc/postgresql/postgresql.conf file. 

So, I don't have a clue what is going on here.

osmeest commented:
Hi,
Is there any reason why this PR is not getting merged into master ?
Looking at the scripts in the branch, I don't see timescaledb extension being included as was originally requested in https://github.com/SynoCommunity/spksrc/issues/3815. Correct ?

hgy59 commented:
> Hi, Is there any reason why this PR is not getting merged into master ? 

There are still issues with cleanup that need to be solved.

> Looking at the scripts in the branch, I don't see timescaledb extension being included as was originally requested in #3815. Correct ?

timescaledb is not related to this PR (it is a different product/project, not a pgsql extension).

DigitalBox98 commented:
It seems the solution consists to deactivate auto vacuum and to create a DSM task to replace it.

If I have some time I will look at it (some useful information are in https://hevodata.com/learn/postgresql-vacuum-command/)

hgy59 committed:
update to v11.22

- update postgresql to v11.22
- update to openssl3

DigitalBox98 commented:
@RMHogervorst 

A few findings on my side : 
- Postgresql integrated to DSM doesn't use the vacuum as the properfy is off (see below screenshot from /etc/postgresql/postgresql.conf, owner is "postgres")
![image](https://github.com/SynoCommunity/spksrc/assets/57635141/f396231f-2b77-49fb-ace9-9bb8ac7b6c81)
- By default, I didn't see any usage for Postgresql 11 and 13 of the vacuum (in /var/packages/postgresql/var/data/postgresql.conf the line is commented #autovacuum = on)

How did you get the vacuum messages in your logs ?

I've tried to launch the vacuum and I didn't get any error : 
/var/packages/postgresql/target/bin/psql -U pgadmin -b postgres -p 5433
vacuum full;

I've actived the vacuum to on, and I will check in the coming days if I see any error on my side.

RMHogervorst commented:
It has been a while since I looked into the logs, I don't remember anymore.

DigitalBox98 commented:
@hgy59 I've got version 13.2 compiling and running fine on my NAS, are you interested in this version ?

RMHogervorst commented:
I'd love to check it out, but it will be a while before I have the time

mreid-tt commented:
Hey @hgy59, I’ve been working on several packages that depend on MySQL/MariaDB, and it got me thinking about potential performance gains with PostgreSQL. I know a lot of developers prefer PostgreSQL for certain applications, and I’m curious to explore this further.

Do you think it would be worth testing PostgreSQL for these packages? In the long run, could it make sense to transition dependencies from MariaDB to PostgreSQL if we find specific advantages? Would love to hear your thoughts on this.

mreid-tt committed:
postgresql: Update to version 17.4

- Upgrade from EOL PostgreSQL 11.22 to current 17.4
- Add cross/libicu dependency (required for PG 17)
- Add LDFLAGS_EX_BE="-Wl,--export-dynamic" to fix undefined symbol
  error in dict_snowball.so during initdb (cross-compile fix)
- Add SERVICE_PORT=5433 for firewall auto-configuration
- Add NO_SERVICE_SHORTCUT=1 (PostgreSQL uses binary protocol, not HTTP)
- Update PLIST for PG 17 binaries and library versions

mreid-tt committed:
postgresql: Modernize package structure

- Unify service-setup.sh for DSM 6 and 7 using run_as_user() helper
- Remove redundant conf.root directory and service-setup.as-root.sh
- Consolidate wizards into single directory for both DSM versions
- Remove wizard.dsm7 directory (redundant)
- Improve wizard security: remove default password, add validation
- Hardcode port 5433 in config (matches SERVICE_PORT, avoids conflicts)
- Disable autovacuum on DSM 7 (non-root user lacks system table permissions)
- Simplify upgrade wizard to informational only
- Add connection instructions to install wizard
- Use SERVICE_PORT variable instead of hardcoded port

mreid-tt committed:
postgresql: Use SERVICE_USER=auto for simpler configuration

- Remove custom conf/privilege file (framework auto-generates)
- Use framework-provided EFF_USER variable (sc-postgresql)
- Add UNSUPPORTED_ARCHS for architectures lacking libicu support
- Fix wizard path from /volume1/@appstore/ to /var/packages/

mreid-tt committed:
postgresql: Improve uninstall wizard and backup validation

- Remove keep/delete data option from wizard (DSM 6 doesn't support it,
  DSM 7 framework handles it automatically)
- Add validation to check backup directory exists and is writable
- Fix chown group to use dynamic lookup instead of hardcoded value

mreid-tt committed:
postgresql: Improve security with scram-sha-256 auth and password complexity

- Use scram-sha-256 authentication instead of trust (requires password)
- Move Unix socket from /tmp to package var directory for persistence
- Add password complexity validation (10+ chars, mixed case, number, special)
- Set default admin username to 'pgadmin'
- Add PostgreSQL's 63-character username limit

mreid-tt committed:
postgresql: Fix authentication for setup and backup

- Use peer auth during initial setup to allow passwordless role creation
- After setup, switch to hybrid auth: system user keeps peer (for backups),
  all other users require scram-sha-256
- Host connections always require scram-sha-256
- Update backup commands to use peer auth (no -h flag)

mreid-tt committed:
Merge branch 'master' into pr/4267

hgy59 commented on spk/postgresql/Makefile:
```diff
@@ -0,0 +1,35 @@
+UNSUPPORTED_ARCHS = $(ARMv5_ARCHS) $(OLD_PPC_ARCHS)
+
+# Use auto-generated service user (sc-postgresql)
+SERVICE_USER = auto
```
This does not work, psql user must not contain `-`

mreid-tt committed:
postgresql: Add PostGIS spatial extension

- Add native/postgresql for pg_config during cross-compilation
- Add cross/libgeos (GEOS 3.12.3)
- Add cross/libproj (PROJ 8.2.1 - compatible with GCC 8)
- Add cross/postgis (PostGIS 3.5.2)
- Auto-create PostGIS extension in template1 and postgres databases

mreid-tt committed:
postgresql: Fix PostGIS cross-compilation for non-x64 architectures

- Restore GEOS 3.12.3 (was downgraded to 3.9.6 causing ARMv7 ODR violations)
- Add pg_config wrapper for cross-compilation paths
- Fix PostGIS configure libpq detection with sed patch
- Exclude comcerto2k (GCC 4.9.3 lacks C++14 for GEOS 3.12.3)
- Require DSM 7.0+ (GCC 4.9 strict linker fails libpq detection)

mreid-tt commented:
hey @hgy59, @DigitalBox98 — I’ve done some work to bring this up to date and ready for merging. It now includes the PostGIS Spatial Extension, which appears to load correctly within the database. I’ve also updated the PR description to reflect the current set of commits.

I’d really appreciate any review and, if possible, some testing of the code and/or the resulting package.

mreid-tt committed:
PostgreSQL: Add SPK_COMMANDS and simplify wizard

hgy59 commented:
> There were already other PRs without PostGIS that supported older DSM versions.

I can't see any merged PR for postgres and there is no postgres package published yet.

hgy59 commented:
> I don’t plan to maintain two separate PostgreSQL packages.

We don't need two separate packages. Only one package, but DSM 6 versions without postgis etc.

mreid-tt committed:
PostgreSQL: Conditionally include PostGIS based on GCC version

- Remove explicit REQUIRED_MIN_DSM and UNSUPPORTED_ARCHS constraints
- Only include postgis dependency when TC_GCC >= 5.0
- Make PostGIS extension creation conditional in service-setup.sh
- Allows building on older DSM versions (DSM 6.2.4) without PostGIS

mreid-tt committed:
PostgreSQL: Add ARMv5 and OLD_PPC_ARCHS to UNSUPPORTED_ARCHS

mreid-tt committed:
Merge branch 'master' into pr/4267

mreid-tt committed:
PostgreSQL: Fix pg_hba.conf alignment
